# Exercise 6.1 - Sourcing Open Data

## Chocolate Bar Ratings - Cleaning and Consistency Checks

## -

### 1. Importing Libraries & Data

In [1]:
# Importing Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create path
path = r'/Users/puneet/Desktop/Chocolate Bar Analysis 09-2025'

In [3]:
# Import Chocolate Bar Ratings CSV dataset
df_cbr = pd.read_csv(os.path.join(path,'02-Data','Original Data Files', 'flavors_of_cacao.csv'), index_col = False)

## -

### 2. Data Cleaning & Consistency Checks

In [4]:
# Check data head
df_cbr.head()

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                   A. Morin                       Agua Grande  1876   
1                   A. Morin                             Kpime  1676   
2                   A. Morin                            Atsane  1676   
3                   A. Morin                             Akata  1680   
4                   A. Morin                            Quilla  1704   

   Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0          2016            63%            France    3.75              
1          2015            70%            France    2.75              
2          2015            70%            France    3.00              
3          2015            70%            France    3.50              
4          2015            70%            France    3.50              

  Broad Bean\nOrigin  
0           Sao Tome  
1               Togo  
2               Togo  
3               Togo  
4               Peru

In [5]:
df_cbr.columns

Index(['Company \n(Maker-if known)', 'Specific Bean Origin\nor Bar Name',
       'REF', 'Review\nDate', 'Cocoa\nPercent', 'Company\nLocation', 'Rating',
       'Bean\nType', 'Broad Bean\nOrigin'],
      dtype='object')

In [6]:
# Rename Columns
df_cbr = df_cbr.rename(columns = {
    'Company \n(Maker-if known)': 'Company',
    'Specific Bean Origin\nor Bar Name': 'Bar_Name',
    'REF': 'Reference_Number',
    'Review\nDate': 'Review_Date',
    'Cocoa\nPercent': 'Cocoa_Percentage',
    'Company\nLocation': 'Company_Location',
    'Bean\nType': 'Bean_Type',
    'Broad Bean\nOrigin': 'Bean_Origin'
})

In [7]:
# Check new column names
df_cbr.columns

Index(['Company \n(Maker-if known)', 'Bar_Name', 'Reference_Number',
       'Review_Date', 'Cocoa_Percentage', 'Company_Location', 'Rating',
       'Bean_Type', 'Bean_Origin'],
      dtype='object')

In [8]:
# Rename Columns (First column name didn't change)
df_cbr.columns = ['Company','Bar_Name', 'Reference_Number',
       'Review_Date', 'Cocoa_Percentage', 'Company_Location', 'Rating',
       'Bean_Type', 'Bean_Origin']

In [9]:
df_cbr.columns

Index(['Company', 'Bar_Name', 'Reference_Number', 'Review_Date',
       'Cocoa_Percentage', 'Company_Location', 'Rating', 'Bean_Type',
       'Bean_Origin'],
      dtype='object')

In [10]:
# Check new table
df_cbr.head()

Company     Bar_Name  Reference_Number  Review_Date Cocoa_Percentage  \
0  A. Morin  Agua Grande              1876         2016              63%   
1  A. Morin        Kpime              1676         2015              70%   
2  A. Morin       Atsane              1676         2015              70%   
3  A. Morin        Akata              1680         2015              70%   
4  A. Morin       Quilla              1704         2015              70%   

  Company_Location  Rating Bean_Type Bean_Origin  
0           France    3.75              Sao Tome  
1           France    2.75                  Togo  
2           France    3.00                  Togo  
3           France    3.50                  Togo  
4           France    3.50                  Peru

In [11]:
# Check data rows and columns
df_cbr.shape

(1795, 9)

- Dataset contains 1795 rows and 9 columns worht of data

In [12]:
# Check info on data columns
df_cbr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1795 non-null   object 
 1   Bar_Name          1795 non-null   object 
 2   Reference_Number  1795 non-null   int64  
 3   Review_Date       1795 non-null   int64  
 4   Cocoa_Percentage  1795 non-null   object 
 5   Company_Location  1795 non-null   object 
 6   Rating            1795 non-null   float64
 7   Bean_Type         1794 non-null   object 
 8   Bean_Origin       1794 non-null   object 
dtypes: float64(1), int64(2), object(6)
memory usage: 126.3+ KB


In [13]:
# Check data types for the columns
df_cbr.dtypes

Company              object
Bar_Name             object
Reference_Number      int64
Review_Date           int64
Cocoa_Percentage     object
Company_Location     object
Rating              float64
Bean_Type            object
Bean_Origin          object
dtype: object

In [14]:
# Change 'Cocoa_Percerntage' column to float, instead of string as it displays percentages. Remove "%" symbol
df_cbr['Cocoa_Percentage'] = df_cbr['Cocoa_Percentage'].str.rstrip('%').astype(float)

In [15]:
# Show new Cocoa_Percentage info
df_cbr['Cocoa_Percentage'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 1795 entries, 0 to 1794
Series name: Cocoa_Percentage
Non-Null Count  Dtype  
--------------  -----  
1795 non-null   float64
dtypes: float64(1)
memory usage: 14.2 KB


In [16]:
df_cbr.head(2)

Company     Bar_Name  Reference_Number  Review_Date  Cocoa_Percentage  \
0  A. Morin  Agua Grande              1876         2016              63.0   
1  A. Morin        Kpime              1676         2015              70.0   

  Company_Location  Rating Bean_Type Bean_Origin  
0           France    3.75              Sao Tome  
1           France    2.75                  Togo

In [17]:
#Check for Mixed-type column
for col in df_cbr.columns.tolist():
  weird = (df_cbr[[col]].map(type) != df_cbr[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cbr[weird]) > 0:
    print (col)

Bean_Type
Bean_Origin


In [18]:
#Change Bean_Type and Bean_Origin to Strings (No numeric values for these columns)
df_cbr['Bean_Type'] = df_cbr['Bean_Type'].astype(str)
df_cbr['Bean_Origin'] = df_cbr['Bean_Origin'].astype(str)

#Check new columns
df_cbr[['Bean_Type','Bean_Origin']].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 2 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Bean_Type    1795 non-null   object
 1   Bean_Origin  1795 non-null   object
dtypes: object(2)
memory usage: 28.2+ KB


In [19]:
#Check for Mixed-type column (After cleaning)
for col in df_cbr.columns.tolist():
  weird = (df_cbr[[col]].map(type) != df_cbr[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_cbr[weird]) > 0:
    print (col)

- No mixed-type columns

In [20]:
#Describe Numerical Data
df_cbr.describe()

Reference_Number  Review_Date  Cocoa_Percentage       Rating
count       1795.000000  1795.000000       1795.000000  1795.000000
mean        1035.904735  2012.325348         71.698329     3.185933
std          552.886365     2.927210          6.323118     0.478062
min            5.000000  2006.000000         42.000000     1.000000
25%          576.000000  2010.000000         70.000000     2.875000
50%         1069.000000  2013.000000         70.000000     3.250000
75%         1502.000000  2015.000000         75.000000     3.500000
max         1952.000000  2017.000000        100.000000     5.000000

- Descriptive Statistics Look Clean

In [21]:
#Check values for 'Bean_Type' (Missing values)
df_cbr['Bean_Type'].value_counts()

Bean_Type
                            887
Trinitario                  419
Criollo                     153
Forastero                    87
Forastero (Nacional)         52
Blend                        41
Criollo, Trinitario          39
Forastero (Arriba)           37
Criollo (Porcelana)          10
Trinitario, Criollo           9
Forastero (Parazinho)         8
Forastero (Arriba) ASS        6
Beniano                       3
Matina                        3
EET                           3
Nacional (Arriba)             3
Criollo, Forastero            2
Amazon, ICS                   2
Trinitario, Forastero         2
Amazon mix                    2
Forastero (Catongo)           2
Nacional                      2
Trinitario (85% Criollo)      2
Criollo (Amarru)              2
Criollo (Ocumare 61)          2
Criollo, +                    1
Forastero (Arriba) ASSS       1
Trinitario, TCGA              1
Trinitario (Amelonado)        1
Trinitario, Nacional          1
nan                           

In [22]:
#Check Null values in Bean_Type column
df_cbr['Bean_Type'].isnull().sum()

np.int64(0)

- *No nulls counted, but blank values showing in table (887 Counts)*

In [23]:
#List Bean Type values to see "Nulls"
list(df_cbr['Bean_Type'])

['\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 'Criollo',
 '\xa0',
 'Criollo',
 'Criollo',
 '\xa0',
 '\xa0',
 'Criollo',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 'Trinitario',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 'Criollo',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 'Forastero (Arriba)',
 'Forastero (Arriba)',
 'Criollo',
 'Trinitario',
 'Criollo',
 'Forastero',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 '\xa0',
 'Forastero (Nacional)',
 'Criollo, Trinitario',
 'Trinitario',
 'Forastero',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 'Trinitario',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 'Trinitario',
 '\xa0',
 '\xa0',
 'Trinitario',
 '\xa0',
 'Criollo',
 'Forastero',
 '\xa0',
 '\xa0',
 'Forastero',
 '\xa0',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 '\xa0',
 'Criollo (Porcelana)',
 'Blend',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 'Trinitario',
 'Trinitario (85%

- We see the value '\xa0' showing up as nulls

In [24]:
#Replace values with 'Unknown'
df_cbr['Bean_Type'] = df_cbr['Bean_Type'].replace('\xa0','Unknown')

In [25]:
#Check 'Bean_Type' Values again (After replacement above)
df_cbr['Bean_Type'].value_counts()

Bean_Type
Unknown                     887
Trinitario                  419
Criollo                     153
Forastero                    87
Forastero (Nacional)         52
Blend                        41
Criollo, Trinitario          39
Forastero (Arriba)           37
Criollo (Porcelana)          10
Trinitario, Criollo           9
Forastero (Parazinho)         8
Forastero (Arriba) ASS        6
Beniano                       3
Matina                        3
EET                           3
Nacional (Arriba)             3
Criollo, Forastero            2
Amazon, ICS                   2
Trinitario, Forastero         2
Amazon mix                    2
Forastero (Catongo)           2
Nacional                      2
Trinitario (85% Criollo)      2
Criollo (Amarru)              2
Criollo (Ocumare 61)          2
Criollo, +                    1
Forastero (Arriba) ASSS       1
Trinitario, TCGA              1
Trinitario (Amelonado)        1
Trinitario, Nacional          1
nan                           

#### Null values have been changed to 'Unknown' in Bean Type Column

In [26]:
#Check 'Bean_Origin' Values
df_cbr['Bean_Origin'].value_counts().head(50)

Bean_Origin
Venezuela                 214
Ecuador                   193
Peru                      165
Madagascar                145
Dominican Republic        141
                           73
Nicaragua                  60
Brazil                     58
Bolivia                    57
Belize                     49
Papua New Guinea           42
Colombia                   40
Vietnam                    38
Costa Rica                 38
Tanzania                   34
Trinidad                   33
Ghana                      33
Mexico                     30
Guatemala                  28
Hawaii                     28
Domincan Republic          25
Jamaica                    20
Grenada                    19
Indonesia                  16
Honduras                   15
Cuba                       11
Sao Tome                   10
Congo                      10
Haiti                       9
Carribean                   8
Uganda                      8
St. Lucia                   8
Fiji                        

- 73 Missing values showing in 'Bean_Origin'

In [27]:
#List Bean Origin values
list(df_cbr['Bean_Origin'])

['Sao Tome',
 'Togo',
 'Togo',
 'Togo',
 'Peru',
 'Venezuela',
 'Cuba',
 'Venezuela',
 'Venezuela',
 'Peru',
 'Panama',
 'Madagascar',
 'Brazil',
 'Ecuador',
 'Colombia',
 'Burma',
 'Papua New Guinea',
 'Venezuela',
 'Peru',
 'Peru',
 'Peru',
 'Bolivia',
 'Peru',
 'Peru',
 'Peru',
 'Fiji',
 'Fiji',
 'Fiji',
 'Fiji',
 'Ecuador',
 'Ecuador',
 'Mexico',
 'Indonesia',
 'Madagascar',
 'Brazil',
 'Trinidad',
 'Vietnam',
 'Madagascar',
 'Venezuela',
 'Peru',
 'Ecuador',
 'Nicaragua',
 'Vietnam',
 'Tanzania',
 'Nicaragua',
 'Peru',
 'Bolivia',
 'Peru',
 'Ecuador',
 'Venezuela',
 'Dominican Republic',
 'Bolivia',
 'Bolivia',
 'Peru',
 'Papua New Guinea',
 'Dominican Republic',
 'Ecuador',
 'Venezuela',
 'Venezuela',
 'Indonesia',
 'Madagascar',
 'Venezuela',
 'Venezuela',
 'Ghana',
 'Ecuador',
 'Ecuador',
 'Ghana',
 'Peru',
 'Peru',
 'Belize',
 'Madagascar',
 'Dominican Republic',
 'Papua New Guinea',
 'Venezuela',
 'Peru',
 'Peru',
 'Venezuela',
 '\xa0',
 'Venezuela',
 'Ecuador',
 'Jamaica',
 

- Replace same "Null" values in Bean_Origin that show \xa0'

In [28]:
#Replace Null Bean_Origin values with 'Unknown'
df_cbr['Bean_Origin'] = df_cbr['Bean_Origin'].replace('\xa0','Unknown')

In [29]:
#Check 'Bean_Origin' Values (After Cleaning)
df_cbr['Bean_Origin'].value_counts().head(10)

Bean_Origin
Venezuela             214
Ecuador               193
Peru                  165
Madagascar            145
Dominican Republic    141
Unknown                73
Nicaragua              60
Brazil                 58
Bolivia                57
Belize                 49
Name: count, dtype: int64

#### Null Values have been changed to 'Unknown' in Bean Origin Column

In [30]:
#Create new dataframe to check for duplicates
df_cbr_dups = df_cbr[df_cbr.duplicated()]

In [31]:
#Check duplicates table
df_cbr_dups

Empty DataFrame
Columns: [Company, Bar_Name, Reference_Number, Review_Date, Cocoa_Percentage, Company_Location, Rating, Bean_Type, Bean_Origin]
Index: []

- No duplicates in data

#### -

- Perform last data checks after cleaning

In [32]:
df_cbr.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1795 entries, 0 to 1794
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Company           1795 non-null   object 
 1   Bar_Name          1795 non-null   object 
 2   Reference_Number  1795 non-null   int64  
 3   Review_Date       1795 non-null   int64  
 4   Cocoa_Percentage  1795 non-null   float64
 5   Company_Location  1795 non-null   object 
 6   Rating            1795 non-null   float64
 7   Bean_Type         1795 non-null   object 
 8   Bean_Origin       1795 non-null   object 
dtypes: float64(2), int64(2), object(5)
memory usage: 126.3+ KB


In [33]:
df_cbr.shape

(1795, 9)

In [34]:
df_cbr.isnull().sum()

Company             0
Bar_Name            0
Reference_Number    0
Review_Date         0
Cocoa_Percentage    0
Company_Location    0
Rating              0
Bean_Type           0
Bean_Origin         0
dtype: int64

In [35]:
df_cbr.dtypes

Company              object
Bar_Name             object
Reference_Number      int64
Review_Date           int64
Cocoa_Percentage    float64
Company_Location     object
Rating              float64
Bean_Type            object
Bean_Origin          object
dtype: object

In [36]:
df_cbr.head(15)

Company        Bar_Name  Reference_Number  Review_Date  Cocoa_Percentage  \
0   A. Morin     Agua Grande              1876         2016              63.0   
1   A. Morin           Kpime              1676         2015              70.0   
2   A. Morin          Atsane              1676         2015              70.0   
3   A. Morin           Akata              1680         2015              70.0   
4   A. Morin          Quilla              1704         2015              70.0   
5   A. Morin        Carenero              1315         2014              70.0   
6   A. Morin            Cuba              1315         2014              70.0   
7   A. Morin    Sur del Lago              1315         2014              70.0   
8   A. Morin  Puerto Cabello              1319         2014              70.0   
9   A. Morin         Pablino              1319         2014              70.0   
10  A. Morin          Panama              1011         2013              70.0   
11  A. Morin      Madagascar              1011         2013              70.0   
12  A. Morin          Brazil              1011         2013              70.0   
13  A. Morin        Equateur              1011         2013              70.0   
14  A. Morin        Colombie              1015         2013              70.0   

   Company_Location  Rating Bean_Type Bean_Origin  
0            France    3.75   Unknown    Sao Tome  
1            France    2.75   Unknown        Togo  
2            France    3.00   Unknown        Togo  
3            France    3.50   Unknown        Togo  
4            France    3.50   Unknown        Peru  
5            France    2.75   Criollo   Venezuela  
6            France    3.50   Unknown        Cuba  
7            France    3.50   Criollo   Venezuela  
8            France    3.75   Criollo   Venezuela  
9            France    4.00   Unknown        Peru  
10           France    2.75   Unknown      Panama  
11           France    3.00   Criollo  Madagascar  
12           France    3.25   Unknown      Brazil  
13           France    3.75   Unknown     Ecuador  
14           France    2.75   Unknown    Colombia

### *Data is clean and consistent*

## - 

## 3. Save and Export New Cleaned Data

In [37]:
#Export cleaned data
df_cbr.to_csv(os.path.join(path, '02-Data', 'Prepared Data Files', 'flavors_of_cacao_cleaned.csv'))